File: backend/app/schemas/workout.py
from datetime import date, datetime
from typing import Optional, List

from pydantic import BaseModel, ConfigDict

from app.schemas.base import ORMModel


# Nested schemas for WorkoutSet
class WorkoutSetCreate(BaseModel):
    exercise_id: str
    set_index: int
    target_sets: Optional[int] = None
    target_reps_min: Optional[int] = None
    target_reps_max: Optional[int] = None
    target_load_type: Optional[str] = None
    target_load_value: Optional[float] = None
    rpe_target: Optional[float] = None
    rest_seconds: Optional[int] = None
    tempo: Optional[str] = None
    is_warmup: bool = False
    notes: Optional[str] = None
    prescription_text: Optional[str] = None


class WorkoutSetRead(ORMModel):
    id: str
    workout_block_id: str
    exercise_id: str
    set_index: int
    target_sets: Optional[int] = None
    target_reps_min: Optional[int] = None
    target_reps_max: Optional[int] = None
    target_load_type: Optional[str] = None
    target_load_value: Optional[float] = None
    rpe_target: Optional[float] = None
    rest_seconds: Optional[int] = None
    tempo: Optional[str] = None
    is_warmup: bool = False
    notes: Optional[str] = None
    prescription_text: Optional[str] = None


class WorkoutSetUpdate(BaseModel):
    id: Optional[str] = None
    set_index: int
    exercise_id: str
    target_sets: Optional[int] = None
    target_reps_min: Optional[int] = None
    target_reps_max: Optional[int] = None
    target_load_type: Optional[str] = None
    target_load_value: Optional[float] = None
    rpe_target: Optional[float] = None
    rest_seconds: Optional[int] = None
    tempo: Optional[str] = None
    is_warmup: bool = False
    notes: Optional[str] = None
    prescription_text: Optional[str] = None


# Nested schemas for WorkoutBlock
class WorkoutBlockCreate(BaseModel):
    block_type: str
    sequence_index: int
    notes: Optional[str] = None
    sets: List[WorkoutSetCreate] = []


class WorkoutBlockRead(ORMModel):
    id: str
    workout_id: str
    block_type: str
    sequence_index: int
    notes: Optional[str] = None
    sets: List[WorkoutSetRead] = []


class WorkoutBlockUpdate(BaseModel):
    id: Optional[str] = None
    sequence_index: int
    block_type: str
    notes: Optional[str] = None
    sets: List[WorkoutSetUpdate] = []


class WorkoutBase(BaseModel):
    title: str
    date: date
    status: str = "draft"  # "draft", "planned", "in_progress", "completed"
    notes: Optional[str] = None
    freeform_log: Optional[str] = None
    client_id: Optional[str] = None
    template_id: Optional[str] = None


class WorkoutCreate(WorkoutBase):
    blocks: List[WorkoutBlockCreate] = []


class WorkoutUpdate(BaseModel):
    title: Optional[str] = None
    date: Optional[str] = None
    status: Optional[str] = None
    notes: Optional[str] = None
    freeform_log: Optional[str] = None
    client_id: Optional[str] = None
    template_id: Optional[str] = None
    blocks: Optional[List[WorkoutBlockUpdate]] = None

    model_config = ConfigDict(from_attributes=True)


class WorkoutRead(ORMModel, WorkoutBase):
    id: str
    trainer_id: str
    created_at: datetime
    updated_at: datetime
    blocks: List[WorkoutBlockRead] = []

File: frontend/types/workout.ts
/**
 * Workout type definitions matching backend schemas
 * These types mirror backend/app/schemas/workout.py and backend/app/models/workout.py
 */

// ============================================================================
// WorkoutSet types
// ============================================================================

export interface WorkoutSetCreate {
  exercise_id: string;
  set_index: number;
  target_sets?: number | null;
  target_reps_min?: number | null;
  target_reps_max?: number | null;
  target_load_type?: string | null;
  target_load_value?: number | null;
  rpe_target?: number | null;
  rest_seconds?: number | null;
  tempo?: string | null;
  is_warmup?: boolean;
  notes?: string | null;
  prescription_text?: string | null;
}

export interface WorkoutSetRead {
  id: string;
  workout_block_id: string;
  exercise_id: string;
  set_index: number;
  target_sets?: number | null;
  target_reps_min?: number | null;
  target_reps_max?: number | null;
  target_load_type?: string | null;
  target_load_value?: number | null;
  rpe_target?: number | null;
  rest_seconds?: number | null;
  tempo?: string | null;
  is_warmup: boolean;
  notes?: string | null;
  prescription_text?: string | null;
}

// ============================================================================
// WorkoutBlock types
// ============================================================================

export interface WorkoutBlockCreate {
  block_type: string;
  sequence_index: number;
  notes?: string | null;
  sets: WorkoutSetCreate[];
}

export interface WorkoutBlockRead {
  id: string;
  workout_id: string;
  block_type: string;
  sequence_index: number;
  notes?: string | null;
  sets: WorkoutSetRead[];
}

// ============================================================================
// Workout types
// ============================================================================

export interface WorkoutBase {
  title: string;
  date: string; // ISO date string (YYYY-MM-DD from backend, or full ISO datetime)
  status: string; // "draft" | "planned" | "in_progress" | "completed"
  notes?: string | null;
  freeform_log?: string | null;
  client_id?: string | null;
  template_id?: string | null;
}

export interface WorkoutCreate extends WorkoutBase {
  blocks?: WorkoutBlockCreate[];
}

export interface WorkoutUpdate {
  title?: string;
  date?: string;
  status?: string;
  notes?: string;
  freeform_log?: string;
  client_id?: string;
  template_id?: string;
}

export interface WorkoutRead extends WorkoutBase {
  id: string;
  trainer_id: string;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
  blocks: WorkoutBlockRead[];
}

// ============================================================================
// UI-specific types for editor state
// ============================================================================

/**
 * Editable workout set for UI state management
 * Includes both server fields and local UI state
 */
export interface EditableWorkoutSet extends WorkoutSetCreate {
  // Optional id for existing sets (used when updating)
  id?: string;
  // UI-only field to track if this is a new set
  _isNew?: boolean;
}

/**
 * Editable workout block for UI state management
 */
export interface EditableWorkoutBlock {
  // Optional id for existing blocks (used when updating)
  id?: string;
  block_type: string;
  sequence_index: number;
  notes?: string | null;
  sets: EditableWorkoutSet[];
  // UI-only field to track if this is a new block
  _isNew?: boolean;
}

/**
 * Full editable workout for UI state management
 */
export interface EditableWorkout extends WorkoutBase {
  id: string;
  trainer_id: string;
  blocks: EditableWorkoutBlock[];
}
